### PR TITLE
PKOv2 announcement post

### PR DESCRIPTION
**Pull request by EronWright**
### Proposed changes

A blog post for the release of Pulumi Kubernetes Operator v2 (beta 1).

### Unreleased product version (optional)

This post pertains to v2.0-beta.1, which we plan to release on Thurs Oct 17.

### Related issues (optional)

Closes https://github.com/pulumi/pulumi-kubernetes-operator/issues/657

### CODE REVIEW

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+Today, we're excited to make [2.0 beta 1](https://github.com/pulumi/pulumi-kubernetes-operator/releases/tag/v2.0-beta.1) of the Pulumi Kubernetes Operator available.
+We've put a new architecture in place to address these limitations and to provide new customization options. Let's dig in!
+
+## What is the Pulumi Kubernetes Operator?
+
+The Pulumi Kubernetes Operator defines a Kubernetes Custom Resource called `pulumi.com/v1/Stack`, which represents a Pulumi [stack](/docs/concepts/stack/). The Pulumi stack can be authored in any supported Pulumi language (TypeScript, Python, Go, .NET, Java, YAML) and can deploy and manage cloud infrastructure in any supported cloud (AWS, Azure, GCP, Kubernetes and 60+ additional cloud and SaaS providers). The Pulumi Kubernetes Operator triggers cloud deployments based on changes to the `Stack` Custom Resource or the resources it uses.
+
+As a result, the Pulumi Kubernetes Operator enables users to specify the desired state of their cloud infrastructure using resources managed directly in their Kubernetes cluster, which trigger creation, update and deletion of the detailed cloud infrastructure they need.
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:18]:
```suggestion
As a result, the Pulumi Kubernetes Operator enables users to specify the desired state of their cloud infrastructure by using resources managed directly in their Kubernetes cluster. Modifying those Kubernetes resources will trigger creation, updates, and deletion of the underlying cloud infrastructure they manage.
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+<!--more-->
+
+For example, the following Kubernetes resource can be created to deploy a Pulumi program which provisions cloud infrastructure to manage NGINX provided in the target Git repo and commit, and future updates to the commit or config settings will deploy updates to the infrastructure.
```
Comment (adamgordonbell) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:22]: I found this sentence a bit hard to parse.

Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:22]:
```suggestion
For example, the Kubernetes Operator can use the Stack resource below to run a Pulumi program, as defined in a GitHub repository, to deploy NGINX. Subsequent commits to the repository, or changes to the Stack's configuration, will cause the Operator to re-deploy the infrastructure.
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+kind: Stack
+metadata:
+  name: nginx-k8s-stack-production
+spec:
+  stack: acmecorp/nginx/production
+  projectRepo: https://github.com/acmecorp/pulumi-nginx
+  commit: 2b0889718d3e63feeb6079ccd5e4488d8601e353
+```
+
+The Pulumi Kubernetes Operator can also be used along with Pulumi’s [Kubernetes provider](/registry/packages/kubernetes/), which allows Pulumi programs to deploy resources to Kubernetes, either in the same cluster as the operator, or in another cluster. This can be used to package up complex Kubernetes infrastructure into more abstracted units, to mix Kubernetes and cloud infrastructure into a single unit of deployment, to define how application workloads are deployed into a cluster, or for a wide variety of additional use cases. The Pulumi Kubernetes provider has full support for the Kubernetes API available in every Pulumi language, including support for [Helm](/blog/full-access-to-helm-features-through-new-helm-release-resource-for-kubernetes/).
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:35]: Not sure if we want to refer to it as "Operator" or "Pulumi Kubernetes Operator" throughout.

```suggestion
The Pulumi Kubernetes Operator can also be used along with Pulumi’s [Kubernetes provider](/registry/packages/kubernetes/), which allows Pulumi programs to deploy resources to Kubernetes -- either in the same cluster as the Operator, or in another cluster. This can be used to package up complex Kubernetes infrastructure into more abstracted units, to mix Kubernetes and cloud infrastructure into a single unit of deployment, to define how application workloads are deployed into a cluster, or for a wide variety of additional use cases. The Pulumi Kubernetes provider has full support for the Kubernetes API available in every Pulumi language, including support for [Helm](/blog/full-access-to-helm-features-through-new-helm-release-resource-for-kubernetes/).
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+## What’s New in Pulumi Kubernetes Operator 2.0?
+
+The 2.0 release is based on a whole new architecture for running Pulumi programs in your Kubernetes cluster.
+The operator now allocates a dedicated pod for each `Stack` to serve as the execution environment for Pulumi stack operations.
+Previously, all stack operations took place within the operator's own pod. This new approach effectively isolates
+each stack's compute and memory resources, improves the isolation of secrets, and opens up new customization options.
+
+These pods are referred to as *workspace pods*.  Under the hood, a new `Workspace` Custom Resource handles the lifecycle
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:44]: Do we want to mention this is modeled after Automation API?

```suggestion
These pods are referred to as *workspace pods* (modeled after Pulumi's [Automation API](https://www.pulumi.com/docs/iac/packages-and-automation/automation-api/concepts-terminology/#workspace)).  Under the hood, a new `Workspace` Custom Resource handles the lifecycle
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+We provide three ways to install the operator:
+
+1. a Pulumi program (see: `deploy/pulumi-operator-yaml/`)
+2. a Helm chart (see: `deploy/helm/pulumi-operator/`)
+3. a simple manifest file (see: `deploy/yaml/install.yaml`)
+
+Please uninstall any earlier version of the Pulumi Kubernetes Operator, then install the new version using one of the above options.
+Note that this will remove any existing `Stack` objects in your cluster, so be sure to export their manifests or have
+a way to restore them.
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:66]: Is there a safer route where they upgrade CRDs instead of destroying them?

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+by the 'restricted' profile of the [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/).
+If necessary, you may configure a stack to use the 'baseline' profile.
+
+A Kubernetes `ServiceAccount` provides an identity for your workspace pod, and we recommend that you create
+a service account for your stack rather than using the `default` account. If your Pulumi program uses the
+Kubernetes provider to manage resources within the cluster, the stack's service account will need permissions, e.g.
+a `RoleBinding` to the `ClusterRole` named `cluster-admin`. See ["Service Accounts"](https://kubernetes.io/docs/concepts/security/service-accounts/) for more information.
+
+The workspace pod has a RPC endpoint that is used by the operator to run stack operations. The pod protects the
+RPC endpoint using Kubernetes RBAC, and for that reason you must bind the `ClusterRole` named `system:auth-delegator` to the
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:80]:
```suggestion
RPC endpoint using [Kubernetes RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/), and for that reason you must bind the `ClusterRole` named `system:auth-delegator` to the
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+  serviceAccountName: my-program
+  ...
+```
+
+### Scalability
+
+Each stack is run in its own "workspace" pod, and so the system scales horizontally to support many stacks.
+The workspace pod is, by default, configured to use minimal resources when idle, and to 'burst' into the unused
+memory and compute resources of the node during a stack operation. You may customize the resource constraints to
+fit the needs of your stack. If your Kubernetes cluster has high utilization, we recommend increasing the resource requests
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:122]:
```suggestion
memory and compute resources of the node during a stack operation. You may customize the resource constraints to
fit the needs of your stack using the `workspaceTemplate` field. If your Kubernetes cluster has high utilization, we recommend increasing the resource requests
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+and limits. See ["Pod Quality of Service Classes"](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/) for more information.
+
+The pod is not discarded at the end of a stack operation, it remains in-place unless the stack is deleted.
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:125]:
```suggestion
The pod is not discarded at the end of a stack operation, instead it remains in-place unless the stack is deleted.
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+We're considering introducing the option to clean up the workspace pod after each operation, to enable you to make a trade-off
+decision between performance and efficiency.
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:127]:
```suggestion
We're considering introducing the option to clean up the workspace pod after each operation to enable you to make a trade-off
decision between performance and efficiency. Please [let us know](https://slack.pulumi.com) if this is something you'd like to see!
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+### Robustness
+
+A Kubernetes cluster is a dynamic environment where a given pod may be terminated at any time, e.g. due to maintenance operations,
+unplanned node events, or resource pressure. This can cause problems for ongoing stack operations, and historically
+may cause a stack to become "locked" (thus necessitating the use of `pulumi cancel`).
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:133]:
```suggestion
unplanned node events, or resource pressure. This can cause problems for ongoing stack operations, and historically
the operator could cause stacks to become "locked", thus necessitating manual intervention with the use of `pulumi cancel`.
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+The `Stack` custom resource continues to be the primary way to use the Pulumi Kubernetes Operator, and your existing
+stack definitions should continue to work after you follow a simple migration procedure:
+
+1. Create a `ServiceAccount` and associated `ClusterRoleBinding` as shown in the Security section.
+2. If your stack uses the [Pulumi Kubernetes Provider](/registry/packages/kubernetes/), create a `RoleBinding` or a `ClusterRoleBinding` to grant the necessary permissions, e.g. to the well-known `ClusterRole` named `cluster-admin`. See ["Using RBAC Authorization"](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles) for more information.
+3. Edit your `Stack` manifest(s) to use the service account, by setting the `spec.serviceAccountName` field.
+4. Double-check that your `Stack` objects do not make use of cross-namespace references.
+
+If you're using a private Git repository as your program source, use the Stack's `secretRefs` field to reference Git credentials.
+SSH-based access to the repository isn't fully supported; contact us to learn more.
```
Comment (blampe) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:232]:
```suggestion
SSH-based access to the repository may require a custom image with your host's public key(s) baked in. Contact us to learn more.
```

**Review by blampe — APPROVED** [review]
This is really great. My feedback is all just optional wordsmithing.

Only question is if we could maybe recommend a safer way to upgrade that doesn't involve removing existing CRDs?

File: content/blog/pulumi-kubernetes-operator-2-0/index.md (hunk 9 above)
Comment (adamgordonbell) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:127]: Perhaps ask for input
```suggestion
The pod stays in place after a stack operation unless the stack is deleted. We're considering an option to clean up the workspace pod after each operation, letting you balance performance and efficiency. Let us know what you think!
```

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
+---
+date: "2024-10-17"
+title: "Pulumi Kubernetes Operator 2.0"
+authors: ["eron-wright"]
+tags: ["Kubernetes", "Continuous-Delivery", "operators"]
+meta_desc: "Pulumi Kubernetes Operator 2.0: Horizontal Scaling, Multi-Tenancy"
+meta_image: operator.png
+---
+
+A few years ago we released the [Pulumi Kubernetes Operator](/blog/pulumi-kubernetes-operator-1-0/), a cloud-native way to manage and deploy cloud infrastructure using Pulumi from within your Kubernetes environment. We've heard your feedback about limitations related to scalability and isolation.
```
Comment (adamgordonbell) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:10]: In the spirit of telling them what you're going to tell them, it may also make sense to mention Robustness and that the changes effect Operability and installation here.

File: content/blog/pulumi-kubernetes-operator-2-0/index.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+the pod logs. The pod's name is based on the stack name; for example, given a stack named `my-stack`, look for
+a pod named `my-stack-workspace-0`. We recommend the [stern](/opt/homebrew/bin/stern) tool to access pod logs.
+See ["Pod and container logs"](https://kubernetes.io/docs/concepts/cluster-administration/logging/#basic-logging-in-kubernetes) for more information.
+
+If you need to run an interactive Pulumi command for your stack, e.g. `pulumi import`, exec into the workspace pod. Navigate to the
+`/share/workspace` directory and you should find your program there.
+
+### Migration
+
+The `Stack` custom resource continues to be the primary way to use the Pulumi Kubernetes Operator, and your existing
```
Comment (adamgordonbell) [inline content/blog/pulumi-kubernetes-operator-2-0/index.md:223]:
```suggestion
Ready to leverage these new features? Let's quickly cover how to migrate your existing setup to Pulumi Kubernetes Operator 2.0. 

The `Stack` custom resource continues to be the primary way to use the Pulumi Kubernetes Operator, and your existing
```

**Review by adamgordonbell — APPROVED** [review]
Overall looks good. Being a bit more explicit about what's new and what is being covered in the beginning would be helpful.

File: content/blog/pulumi-kubernetes-operator-2-0/index.md (hunk 4 above)
Comment (EronWright) [reply content/blog/pulumi-kubernetes-operator-2-0/index.md:44]: No; I got feedback from Luke to the effect that the connection should be downplayed.

File: content/blog/pulumi-kubernetes-operator-2-0/index.md (hunk 5 above)
Comment (EronWright) [reply content/blog/pulumi-kubernetes-operator-2-0/index.md:66]: I wholeheartedly agree but the effort is beyond the scope of the beta.
